experiments/experiment_prototype.py

```diff
@@ -1,385 +1,1246 @@
+        """
+        Before adding the slice, ensure that the internal parameters are set, remove unnecessary
+        keys and check values of keys that are needed, and set defaults of keys that are optional.
+        
+        The following are always able to be defaulted, so are optional:
+        "txantennas", "rx_main_antennas", "rx_int_antennas", "pulse_shift", "scanboundflag", 
+        "scanbound", "acf", "xcf", "acfint", "wavetype", "seqtimer"
+
+
+        The following are always required for processing acf, xcf, and acfint which we will assume
+        we are always doing:
+        "pulse_sequence", "mpinc", "pulse_len", "nrang", "frang", "intt", "intn", "beam_angle",
+        "beam_order"
+
+        The following are required depending on slice type:
+        "txfreq", "rxfreq", "clrfrqrange" 
+        
+        :param: exp_slice: a slice to setup
+        :return: complete_slice : a checked slice with all defaults
+        """
+
+        complete_slice = copy.deepcopy(exp_slice)
+
+        # None values are useless to us - if they do not exist we know they are None.
+        for key, value in complete_slice.items():
+            if value is None:
+                complete_slice.pop(key)
+
+        try:
+            assert 'rxfreq' in complete_slice.keys() or 'txfreq' in complete_slice.keys() \
+                   or 'clrfrqrange' in complete_slice.keys()
+        except AssertionError:
+            errmsg = 'An rxfreq, txfreq, or clrfrqrange must be specified in a slice'
+            raise ExperimentException(errmsg, exp_slice)
+
+        self.set_slice_identifiers(complete_slice)
+        self.check_slice_specific_requirements(complete_slice)
+        self.check_slice_minimum_requirements(complete_slice)
+        complete_slice = self.set_slice_defaults(complete_slice)
+        # set_slice_defaults will check for any missing values that should be given a default and
+        # fill them.
+
+        errors = self.check_slice(complete_slice)
+
+        if errors:
+            raise ExperimentException(errors)
+
+        return complete_slice
+
+    def self_check(self):
         """
         Check that the values in this experiment are valid
         """
 
-        if (self.cponum<1):
-            errmsg="Error: No objects in control program"
-            sys.exit(errmsg) # REVIEW 6 Add a todo for error handling. Perhaps use exceptions instead.
+        if self.num_slices < 1:
+            errmsg = "Error: Invalid num_slices less than 1"
+            raise ExperimentException(errmsg)
 
-        #TODO: somehow check if self.cpid is not unique
+        # TODO: somehow check if self.cpid is not unique - incorporate known cpids from git repo?
 
-        for cpo in range(self.cponum):
-            selferrs=selfcheck_piece(self.cpo_list[cpo],self.config)
-            if (not selferrs): # REVIEW #39 unneeded parenthesis
+        for a_slice in self.slice_ids:
+            selferrs = self.check_slice(self.slice_dict[a_slice])
```

Comment:
I guess I did this so it could be called by build_scans (which is called in experiment_handler before running the experiment) because I want to protect it from the user changing something improperly in their experiment (by not using add_slice, edit_slice, or del_slice). I think this is somewhat preventable by giving the front underscores on __slice_dict, but that doesn't actual make it a private variable in python just hides it a bit.